### PR TITLE
Switch Elasticsearch IT tests to use testcontainers

### PR DESCRIPTION
**Pull request by bernd**
This PR switches all Elasticsearch `*IT.java` tests over to use the Docker based [testcontainers](https://www.testcontainers.org/) project.

Previously we have been using `elasticsearch-maven-plugin` and a custom fork of `nosqlunit-elasticsearch-http` to write and run the IT tests for Elasticsearch. That made it hard to run the tests in development because a developer needed to start the ES instance manually. The custom fork of the `nosqlunit-elasticsearch-http` library also doesn't support ES 6.x yet and would need an update.

One major change in how we execute the tests is, that we now start a new Elasticsearch instance for every test that is running. That means the tests became pretty slow compared to the previous way of running the tests. There are some options to improve that situation in the future (e.g. start a single instance for every test class) but I would like to do that later to avoid making this PR even bigger.

Summary of benefits from this PR:

- Easier to support new Elasticsearch versions because no changes to test libraries needed
- Getting rid of a custom library fork we would have to maintain
- Easier to run IT tests in development because no manual ES start needed
- Removal of lots of indirections through the nosqlunit code makes test classes and data loading easier to understand

Summary of drawbacks from this PR:

- Slower test runs (for now)
- Adds dependency on Docker

### CODE REVIEW

**Comment by mpfz0r** [comment]
Works fine for me :+1: 
One test takes about 13 seconds.

**Comment by alex-konn** [comment]
> Works fine for me +1
> One test takes about 13 seconds.

After implementing reuse of the ES instance execution time per test is way down to less than a second for most of them.

File: graylog2-server/src/test/java/org/graylog/testing/elasticsearch/Client.java [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+        waitForGreenStatus(indexName);
+
+        return indexName;
+    }
+
+    public void deleteIndices(String... indices) {
+        for (String index : indices)
+            if (indicesExists(index)) {
+                final DeleteIndex deleteIndex = new DeleteIndex.Builder(index).build();
+                executeWithExpectedSuccess(deleteIndex, "fauled to delete index " + index);
```
Comment (bernd) [inline graylog2-server/src/test/java/org/graylog/testing/elasticsearch/Client.java]: Small typo:
```suggestion
                executeWithExpectedSuccess(deleteIndex, "failed to delete index " + index);
```

File: graylog2-server/src/test/java/org/graylog/testing/elasticsearch/ElasticsearchBaseTest.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    private void addGraylogDefaultIndexTemplate() {
+        addIndexTemplates(getGraylogDefaultMessageTemplates(elasticsearch.version()));
+    }
+
+    private static Map<String, Map<String, Object>> getGraylogDefaultMessageTemplates(Version version) {
+        return Collections.singletonMap("graylog-test-internal", indexMapping(version).messageTemplate("*", "standard", -1));
+    }
+
+    private static IndexMapping indexMapping(Version elasticsearchVersion) {
+        switch (elasticsearchVersion.getMajorVersion()) {
```
Comment (bernd) [inline graylog2-server/src/test/java/org/graylog/testing/elasticsearch/ElasticsearchBaseTest.java]: We can probably replace this code by using `IndexMappingFactory` which has support for ES 6.

File: graylog2-server/src/test/java/org/graylog/testing/elasticsearch/ElasticsearchInstance.java [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+    private static final Logger LOG = LoggerFactory.getLogger(ElasticsearchInstance.class);
+
+    private static final Map<Version, ElasticsearchContainer> containersByVersion = new HashMap<>();
+
+    private static final String PROPERTIES_RESOURCE_NAME = "elasticsearch.properties";
+
+    private static final String DEFAULT_IMAGE_OSS = "docker.elastic.co/elasticsearch/elasticsearch-oss";
+    private static final String DEFAULT_IMAGE = "elasticsearch";
+    private static final String DEFAULT_VERSION = "6.5.2";
```
Comment (bernd) [inline graylog2-server/src/test/java/org/graylog/testing/elasticsearch/ElasticsearchInstance.java]: Let's use the latest 6.x version here.

File: graylog2-server/src/test/java/org/graylog/testing/elasticsearch/ElasticsearchInstance.java [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        if (!containersByVersion.containsKey(version)) {
+            containersByVersion.put(version, startNewContainerInstance(image));
+        }
+        return containersByVersion.get(version);
+    }
+
+    private static ElasticsearchContainer startNewContainerInstance(String image) {
+        final ElasticsearchContainer container = new ElasticsearchContainer(image)
+                .withEnv("ES_JAVA_OPTS", "-Xms512m -Xmx512m")
+                .withEnv("discovery.type", "single-node")
```
Comment (bernd) [inline graylog2-server/src/test/java/org/graylog/testing/elasticsearch/ElasticsearchInstance.java]: Does this actually work as an environment variable? I only used the `discovery.type` and `action.auto_create_index` settings as program arguments via `./bin/elasticsearch -E action.auto_create_index`.

File: graylog2-server/src/test/java/org/graylog/testing/elasticsearch/ElasticsearchInstance.java (hunk 3 above)
Comment (alex-konn) [reply graylog2-server/src/test/java/org/graylog/testing/elasticsearch/ElasticsearchInstance.java]: Changed to 6.8.4

File: graylog2-server/src/test/java/org/graylog/testing/elasticsearch/ElasticsearchInstance.java (hunk 4 above)
Comment (alex-konn) [reply graylog2-server/src/test/java/org/graylog/testing/elasticsearch/ElasticsearchInstance.java]: As of the docs [here](https://www.elastic.co/guide/en/elasticsearch/reference/6.8/docker.html#docker-configuration-methods) it does

File: pom.xml [hunk 5]
```diff
@@ -165,9 +165,9 @@
         <junit.version>4.12</junit.version>
         <mockito.version>2.23.4</mockito.version>
         <nosqlunit.version>1.0.0-rc.5</nosqlunit.version>
-        <nosqlunit-elasticsearch-http.version>1.0.0-4+jackson</nosqlunit-elasticsearch-http.version>
         <restassured.version>2.9.0</restassured.version>
         <system-rules.version>1.18.0</system-rules.version>
+        <testcontainers.version>1.10.2</testcontainers.version>
```
[the remark this answers is not in the document]
Comment (alex-konn) [reply pom.xml]: done

File: graylog2-server/src/test/java/org/graylog/testing/elasticsearch/Client.java (hunk 1 above)
Comment (alex-konn) [reply graylog2-server/src/test/java/org/graylog/testing/elasticsearch/Client.java]: done

File: graylog2-server/src/test/java/org/graylog/testing/elasticsearch/ElasticsearchBaseTest.java (hunk 2 above)
Comment (alex-konn) [reply graylog2-server/src/test/java/org/graylog/testing/elasticsearch/ElasticsearchBaseTest.java]: done

**Comment by bernd** [comment]
@alex-konn Feel free to approve and merge. I cannot approve my own pull request. :joy:

**Pull request merged**